Use inventory for automatic component registration

jamiebrynes7 opened the pull request:
This PR introduces the use of the inventory crate for automatic component registration. 

- Added `WrappedVTable` as you must own the type that you call `inventory::collect!()` on. 
	- Could look into whether a typedef is sufficient for this.
- Driveby fix on the worker configs (FPL changes) 

Fixes #57 
Fixes #71 (most of the work appeared to have been done for this..?)

randomPoison commented on project-example/src/main.rs:
```diff
@@ -24,16 +24,8 @@ mod generated;
-        .add_component::<improbable::Interest>()
-        .add_component::<improbable::Position>();
+
+    let components = ComponentDatabase::new();
```
🎉✨💖✨🎉

davedissian replied:
Lets delete this line too, and just roll it into the connection object.

randomPoison commented on spatialos-sdk/src/worker/component.rs:
```diff
@@ -271,8 +274,20 @@ pub(crate) unsafe fn handle_copy<T>(handle: *mut raw::c_void) -> *mut raw::c_voi
+}
+
+impl WrappedVTable {
+    pub fn new(vtable: Worker_ComponentVtable) -> Self {
```
Rather than taking the C type as a parameter, I'd suggest having `new` take the component type as a type parameter and then call `create_component_vtable` internally. This simplifies the public interface, and allows us to keep `create_component_vtable` a private implementation detail.

davedissian replied:
Agreed. Could probably just roll `create_component_vtable` directly into `new` as well.

jamiebrynes7 replied:
That's a good shout

randomPoison commented on spatialos-sdk/src/worker/component.rs:
```diff
@@ -271,8 +274,20 @@ pub(crate) unsafe fn handle_copy<T>(handle: *mut raw::c_void) -> *mut raw::c_voi
+
 // Vtable implementation functions.
-fn create_component_vtable<C: Component>() -> Worker_ComponentVtable {
+pub fn create_component_vtable<C: Component>() -> Worker_ComponentVtable {
```
As noted in the other comment, I think it would be better to keep this private and have `VTable::new` call it internally.

davedissian replied:
+1

jamiebrynes7 replied:
Yeah that makes sense 👍

davedissian commented on spatialos-sdk/src/worker/component.rs:
```diff
@@ -237,14 +239,15 @@ pub struct ComponentDatabase {
 impl ComponentDatabase {
     pub fn new() -> Self {
+        let mut vtables = Vec::new();
+        for table in inventory::iter::<WrappedVTable> {
```
How does this work? Is it reflection done at runtime, or is the inventory something generated at compile time? This is really really neat.

jamiebrynes7 replied:
From https://github.com/dtolnay/inventory: 

> Inventory is built on the ctor crate which provides module initialization functions for Rust similar to __attribute__((constructor)) in C. Each call to inventory::submit! produces a shim that evaluates the given expression and registers it into a registry of its corresponding type. This registration happens dynamically as part of life-before-main for statically linked elements. Elements brought in by a dynamically loaded library are registered at the time that dlopen occurs.

davedissian commented on spatialos-sdk/src/worker/component.rs:
```diff
@@ -271,8 +274,20 @@ pub(crate) unsafe fn handle_copy<T>(handle: *mut raw::c_void) -> *mut raw::c_voi
     Arc::into_raw(copy) as *mut _
 }
 
+pub struct WrappedVTable {
```
Agreed.

jamiebrynes7 replied:
Yup - good call

davedissian commented on spatialos-sdk/src/worker/component.rs:
```diff
@@ -237,14 +239,15 @@ pub struct ComponentDatabase {
 
 impl ComponentDatabase {
```
I think the ComponentDatabase type is completely redundant at this point. Might be worth moving the code that generates a vtable array into a private free standing function that is called by the connection/snapshot `new` functions when generating a vtable to pass to the C API.

jamiebrynes7 replied:
Yeah I was having a similar thought - was looking into if we can just expose a single static `ComponentDatabase` instance that any type that requires it could reference, but `lazy_static` and `* c_void` are not compatible. 

Equally, the cost of alloc'ing a vtable each time its required isn't likely to be particularly high

davedissian replied:
Yeah, a vtable is just a 16 function pointers for each component

randomPoison replied:
We could probably switch to having an optional flag on `WorkerConnection` like `use_vtable_serialization` that enables vtable registration, and then gather the vtables internally.

jamiebrynes7 commented:
Alright @davedissian, @randomPoison I think I've addressed the comments above: 

- Renamed `WrappedVTable` to `VTable`.
- Pulled `create_component_vtable<C: Component>()` into `VTable`.
- Made `ComponentDatabase` internal to the crate
- Added `generate_component_vtable()` method to `ConnectionParameters` to opt into the vtable (on second thought - it may be worth having this behaviour be default and have a `remove_vtables()` method if a user wants to use the schema types directly.
- Re-exported `inventory` in `worker::component` so the consumer of the generated code doesn't need to add `inventory` to their Cargo.toml and thus leak implementation details

davedissian commented on project-example/src/connection_handler.rs:
```diff
@@ -93,9 +94,10 @@ pub fn get_connection(opt: Opt, components: ComponentDatabase) -> Result<WorkerC
                     .using_tcp()
-                    .using_external_ip(true),
+                    .using_external_ip(true)
+                    .generate_component_vtable(),
```
Nit: Would prefer naming this to something that isn't "vtable", as that's an implementation detail. Maybe `enable_internal_serialization()`

jamiebrynes7 replied:
Yup - seems sensible, will change
